client/allocrunner/taskrunner/task_runner.go

```diff
@@ -394,6 +394,22 @@ func (tr *TaskRunner) Run() {
 	defer close(tr.waitCh)
 	var result *drivers.ExitResult
 
+	tr.stateLock.RLock()
+	dead := tr.state.State == structs.TaskStateDead
+	tr.stateLock.RUnlock()
+
+	// if restoring a dead task, ensure that task is cleared and all post hooks
+	// are called without additional state updates
+	if dead {
+		// clear driver handle if it was successfully restored on
+		// already dead task
+		tr.clearDriverHandle()
+		if err := tr.stop(); err != nil {
+			tr.logger.Error("stop failed on terminal task", "error", err)
+		}
+		return
```

Comment:
I think we may also want to call `tr.TaskStateUpdated()` since task states are persisted _before_ the AR is notified. Therefore I think the following could happen:

1. 2 tasks in an alloc start: a leader service, and a sidecar
2. Leader task exits, persists TaskStateDead
3. agent crashes before TaskStateUpdated is called
4. agent restarts, returns here due to TaskStateDead

At this point _I do not think_ anything will have told the sidecar service to exit despite its leader dying. If you call `TaskStateUpdated` here, then all of the leader died detection logic in AR will be run: https://github.com/hashicorp/nomad/blob/master/client/allocrunner/alloc_runner.go#L415-L438

This could be done in a followup PR as well since I think your changes improve the situation.